launch/launch/substitutions/equals_substitution.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the EqualsSubstitution substitution."""
+
+import math
+
+from typing import Any
+from typing import Iterable
+from typing import Optional
+from typing import Text
+from typing import Union
+
+from ..frontend import expose_substitution
+from ..launch_context import LaunchContext
+from ..some_substitutions_type import is_some_substitutions_type
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution
+from ..utilities import normalize_to_list_of_substitutions
+from ..utilities.type_utils import perform_substitutions
+
+
+def _str_is_bool(input: Text) -> bool:
+    """Check if string input is convertible to a boolean."""
+    if not isinstance(input, Text):
+        return False
+    else:
+        return input.lower() in ('true', 'false', '1', '0')
+
+
+def _str_is_float(input: Text) -> bool:
+    """Check if string input is convertible to a float."""
+    try:
+        float(input)
+        return True
+    except ValueError:
+        return False
+
+
+@expose_substitution('equals')
+class EqualsSubstitution(Substitution):
+    """
+    Substitution that checks if two inputs are equal.
+
+    Returns 'true' or 'false' strings depending on the result.
+    """
+
+    def __init__(
+        self,
+        left: Optional[Union[Any, Iterable[Any]]],
+        right: Optional[Union[Any, Iterable[Any]]]
+    ) -> None:
+        """Create an EqualsSubstitution substitution."""
+        super().__init__()
+
+        if is_some_substitutions_type(left):
+            self.__left = normalize_to_list_of_substitutions(left)
+        elif left is None:
+            self.__left = ""
+        elif isinstance(left, bool):
+            self.__left = str(left).lower()
+        else:
+            self.__left = str(left)
```

Comment:
I think we can simplify the implementation if we convert the arguments here to [`TextSubstitution`s](https://github.com/ros2/launch/blob/76ade26d3f7bcb6a5a1b72a512a8c098e6d53f04/launch/launch/substitutions/text_substitution.py#L23) and then call `normalize_to_list_of_substitutions`. Then, we don't need to check if the arguments are substitutions inside of `perform`.

Reply:
https://github.com/ros2/launch/pull/649/commits/bec67457dbb28d6a81e5ce3964337c082505a81e !